docs/design/2025-03-17-sys-session-manage-enhancement.md

```diff
@@ -0,0 +1,173 @@
+        if success {
+            p.Put(se)
+        } else {
+            se.Close()
+        }
+    }()
+
+    if err = fn(se); err != nil {
+        return err
+    }
+    success = true
+    return nil
+}
+```
+
+We can see that the `WithSession` method always calls `p.put(se)` within a defer function to ensure the session is returned to the pool. Some exceptions are when a panic or error happens. In such cases, the session is destroyed instead of being returned to the pool, as it may be in an undefined state, making it safer to discard rather than reuse.
+
+The `Pool` should be responsible for managing the internal sessions' registration and unregistering, and the callers should not care about it. These operations are handled by `Pool.get` and `Pool.put` methods:
+
+```go
+func (p *Pool) get() (*Session, error) {
+    // get a session from the pool
+    se, err := getFromPool()
+    if err != nil {
+        return nil, err
+    }
+    // register the internal session to the map
+    infosync.StoreInternalSession(se.GetSessionContext())
+    return se, nil
+}
+
+func (p *Pool) put(s *Session) {
+    // rollback the session first to reset the internal state.
+    s.Rollback()
+    // unregister the internal session from the map
+    infosync.DeleteInternalSession(s.GetSessionContext())
+    // put the session back to the pool
+    p.putToPool(s)
+}
+```
+
+The above code has been simplified for demonstration purposes. The session pool stores the internal session in a map when `Pool.get` is called and removes it from the map in `Pool.put`. Additionally, before being returned to the pool, the session is automatically rolled back to ensure its internal state is reset.
+
+With the newly implemented session pool that manages the session lifecycle, developers no longer need to worry about the session's state or lifecycle. 
+
+### Session
+
+The type `Session` is not a "real" one but a wrapper of the internal session, see:
+
+```go
+// Session is public to callers.
+type Session struct {
+    internal *session
+}
+```
+
+It provides a set of methods that proxy access to the internal session. When the caller invokes `WithSession`, a new `Session` instance is created and passed to the function. Once the function completes, the `Session` instance is destroyed, and only the internal session is returned to the pool.
+
+We have several reasons for proposing the above design:
+
+- The wrapped methods allow for additional checks to prevent misuse of the internal session. For example, if a caller attempts to invoke methods after the session has been returned to the pool, we can detect this and return an error.
+- The `Session` type can be extended with new methods, providing more convenient functionality for internal use.
+- The original session contains numerous methods, many of which may not be necessary for the caller. By filtering out unnecessary methods, we can expose only the required ones, simplifying the interface and improving usability.
```

Comment:
> may not be necessary for the caller

to achieve this goal, we can check and change if `session`'s methods are exposed. So it's not very accurate that it's done by `Session`

Reply:
`session` is also used by `Pool`, and some methods are provided to `Pool` only. One way is to make these pool-only methods private, but I don't want to do it. I think the current design is more clear:
- `session` is provided to both `Session` and  `Pool`, it wraps the internal `sessionctx.Context` and provide a safe access. 
- `Pool` and `Session` should only used the public method of `session` to make a clear semantic.
- `Session` is only used to proxies the caller's access to `session`, expose necessary methods only.